lib/ecmascript.ts

```diff
@@ -2266,7 +2266,7 @@ export const ES = ObjectAssign({}, ES2020, {
     return ES.BalanceISODateTime(year, month, day, hour, minute, second, millisecond, microsecond, nanosecond);
   },
   GetIANATimeZoneNextTransition: (epochNanoseconds, id) => {
-    const uppercap = ES.SystemUTCEpochNanoSeconds() + 366 * DAYMILLIS * 1e6;
+    const uppercap = ES.SystemUTCEpochNanoSeconds().plus(366).multiply(DAYMILLIS).multiply(1e6);
```

Comment:
Oops, approved too soon. This new code isn't accurate because of order of operations and precedence. 

The original code adds 366 days' worth of nanoseconds. 

The new code adds 366 nanoseconds, then multiplies the new number of nanoseconds by 86400000 and then multiplies again by 10^6. That's a totally different operation. 

AFAIK, 366 days is too large a number of nanoseconds to fit inside MAX_SAFE_INTEGER. (~104 days is the limit, if I remember correctly).  So the easy alternative of `plus(366 * DAYMILLIS * 1e6)` won't work. 

One  suggestion would be to define a bigint constant for 366 days of nanoseconds, and then add that constant in the method above.

Reply:
Also, I'm curious about test coverage. Is there a test that calls this function? If so, how long does it take to run?  I'd assume that running through the `while` loop below this code an additional 8.64 trillion times might have taken a loooong time, but I didn't see the tests taking much longer. Which made me wonder whether this was being tested properly. (Where "properly" implies using a few dates that are far enough from epoch that the test should be really slow.)

Reply:
Looking over the implementation of that while loop, I don't think that accidentally creating a very large upper cap makes a huge difference for the current tests as it seems to break out when it detects the transition point, and jumps forward in 2 week increments before breaking out of that loop and bisecting that 2 week chunk to get the actual transition point. That sounds relatively quick to me assuming that the upper cap really is ~1 year - that's only 26 iterations of the loop.

I suppose to test this there needs to be a timezone without any transition points within the upper cap of 366 days after the start point. Is that a thing?

Reply:
Early morning coffee brain thought: UTC doesn't have transitions, right?

Reply:
I've looked at this code some more, and now I just have more questions than I started with:
 - why is the upper cap defined based on the current system time + 266 days, instead of the passed in timestamp + 366 days?
  - why does the while loop compare leftNanos, which is (mostly) static and doesn't change per-loop invocation, instead of rightNanos, which is the time being advanced forward by 2 weeks per iteration?

Reply:
Never mind - one of the testcases explains the first bullet point. Presumably, it checks up to ~the next year from the current time as that's likely how far in advance the transitions are defined in system/browser data, and we do want to support finding transitions that were newly defined more than a year from the given start point in the future.

Reply:
... and now the rest of the code makes sense as well. The left and right nanos values are the sliding window which later gets bisected. Comparing leftNanos against the cap makes sense given that it is also moving.

Clearly I need to go get another coffee... Sorry for all the notifications!

Reply:
Not sure it's related to this order-of-operations bug, but the cookbook tests break for the equivalent PR over in https://github.com/tc39/proposal-temporal/pull/1848, so we may want to see why and turn that "why" into a Demitasse test in both places.

> Looking over the implementation of that while loop, I don't think that accidentally creating a very large upper cap makes a huge difference for the current tests

Yep, because we must only be testing only zones with transitions (which won't show the problem) or offset time zones (e.g. `-08:00`) or `UTC` which are special cased and won't run into the bug either. 

> I suppose to test this there needs to be a timezone without any transition points within the upper cap of 366 days after the start point. Is that a thing?

Correct. `Etc/GMT+10` or `Asia/Calcutta` are good candidates to try.

You'll know you've found the right zone when the current code in this PR runs for ~1.17 × 10<sup>17</sup> iterations. 😄  With current computers that's only about a million years. Better get on the AWS Quantum&trade; beta!

> Early morning coffee brain thought: UTC doesn't have transitions, right?

It has no transitions, but UTC is special-cased in the implementation. UTC is still a good test (of the special-casing) but we also need to test another TZ without transitions that isn't special-cased.

Reply:
BTW, I opened #60 to deal with problems discussed above. We probably don't want the GetNextTransition / GetPreviousTransition to be such an easy DoS attack.

Reply:
Also I opened #59 to see if we can run cookbook tests in this repo just like we run Test262.